### ISSUE TITLE
Version conflicts installing lock file when multiple git dependencies rely on another git dependency

### ISSUE THREAD

**Issue by conorvtx** [astral-sh/uv#6041]
When running

```
uv pip install -r requirements/lock.txt
```

I get something like

```
error: Requirements contain conflicting URLs for package `repo1`:
- git+ssh://git@github.com/org/repo1.git
- git+ssh://git@github.com/org/repo1.git@f6f1bd2f14f887fe63b15632d3a18109ab7a8105
```

The lock file is generated with

```
uv pip compile -q --generate-hashes pyproject.toml -o requirements/lock.txt
```

and has

```
repo2 @ git+ssh://git@github.com/org/repo2.git@f6f1bd2f14f887fe63b15632d3a18109ab7a8106
    # via current_project (pyproject.toml)

repo3 @ git+ssh://git@github.com/org/repo3.git@f6f1bd2f14f887fe63b15632d3a18109ab7a8104
    # via repo2

repo1 @ git+ssh://git@github.com/org/repo1.git@f6f1bd2f14f887fe63b15632d3a18109ab7a8105
    # via
    #   current_project (pyproject.toml)
    #   repo2
    #   repo3
```

The pyproject.toml file for current_project contains dependencies like the following

```toml
dependencies = [
    "repo1 @ git+ssh://git@github.com/org/repo1.git",
    "repo2 @ git+ssh://git@github.com/org/repo2.git",
]
```

repo2 and repo3 use `setup.py`s that have repo1 as a dependency. They require it as follows

```python
# setup.py

def read_requirements(path):
    return [
        line.strip()
        for line in read(path).split("\n")
        if not line.startswith(('"', "#", "-"))
    ]


setup(
    ...
    install_requires=read_requirements("requirements.txt"),
    ...
)
```

```
# requirements.txt

repo1 @ git+ssh://git@github.com/org/repo1.git
```

---

I am surprised I get the error

```
error: Requirements contain conflicting URLs for package `repo1`:
- git+ssh://git@github.com/org/repo1.git
- git+ssh://git@github.com/org/repo1.git@f6f1bd2f14f887fe63b15632d3a18109ab7a8105
```

I would expect `git+ssh://git@github.com/org/repo1.git@f6f1bd2f14f887fe63b15632d3a18109ab7a8105` to satisfy the more general constraint of `git+ssh://git@github.com/org/repo1.git`.

Currently to get around this, I have to update the lock file in `current_project` every time `repo1` has a new commit on the main branch.

**Comment by conorvtx** [comment]
I am surprised that this fails at the install step rather than at the compile step. I would expect that this problem would be surfaced when making the lock file rather than using the lock file. I thought that the lock file would have all the information needed to rebuild a dependency graph as it has the pinned commits of any package installed through git.

**Comment by charliermarsh** [comment]
I believe that error is occurring during the resolver step.

**Comment by charliermarsh** [comment]
Perhaps you intend to be using `uv pip sync`? `uv pip install` does a resolution of the input requirements.

**Comment by charliermarsh** [comment]
Honestly that will probably fail too though.

**Comment by conorvtx** [comment]
actually `uv pip sync` works! however, i was originally not using `sync` because it would remove packages installed by conda in the conda environment.

**Comment by charliermarsh** [comment]
I think what you're trying to do here should probably work (but agree it doesn't).

**Comment by NickCrews** [comment]
I just ran into this (or a similar flavor) too. IDK if this is significant, but in my case the version is `main` for both, so it should be even more obvious that the versions are compatible?

```
error: Requirements contain conflicting URLs for package `noatak`:
- git+https://github.com/ShipCreekGroup/noatak.git@main
- git+https://github.com/ShipCreekGroup/noatak.git@main
```

**Comment by charliermarsh** [comment]
Can you come up with a reproduction that I can use to debug? That’s a confusing failure.

**Comment by NickCrews** [comment]
Hmm, after trying to make out a repro, I think I was running into an issue because one of the remote deps was stale, when I pushed there the error went away. I think ignore me. If it's useful for anyone else, I made a tiny [repo here](https://github.com/NickCrews/uv-bug-6041) trying to repro, I think that might be a nice template for someone trying to see how multiple libs and transitive deps might interact

**Comment by charliermarsh** [comment]
Thank you.

**Comment by jasongill** [comment]
Correct, sorry for the mix-up - I saw the ticket open for "Requirements contain conflicting URLs for package...." and figured it was the same issue

**Comment by charliermarsh** [comment]
No problem. I will look at it -- feel free to create a separate issue.

**Comment by ghidalgo3** [comment]
I ran into this problem too, and found that it reproduces when you:
1. Have a `pyproject.toml` that references multiple `git+` dependencies (for example `A` and `B`)
2. There are internal dependencies between them (`A` declares `B` as a dependency).

The solution is to only reference the top-level one. So in my example, if package `A` depends on `B`, then it's not necessary to list `B` in the list of dependencies of the `pyproject.toml`.

For reference, this is the `dependencies` table I've been working on in an internal project:

```toml
dependencies = [
    # "pctasks.core @ git+https://github.com/microsoft/planetary-computer-tasks@main#subdirectory=pctasks/core",
    # "pctasks.task @ git+https://github.com/microsoft/planetary-computer-tasks@main#subdirectory=pctasks/task",
    # "pctasks.cli @ git+https://github.com/microsoft/planetary-computer-tasks@main#subdirectory=pctasks/cli",
    # "pctasks.client @ git+https://github.com/microsoft/planetary-computer-tasks@main#subdirectory=pctasks/client",
    "pctasks.run @ git+https://github.com/microsoft/planetary-computer-tasks@main#subdirectory=pctasks/run",
]
```

`pctasks.run` indirectly references the rest of the projects of `pctasks`, but if I add them explicitly as dependencies then I will get an error like:

```
    Updated https://github.com/microsoft/planetary-computer-tasks (a27f47c9413faea76117b205d6da8a86a0bb8d9b)
error: Requirements contain conflicting URLs for package `pctasks-client`:
- file:///private/var/folders/p0/vbz6hyr901z17__1xr798h2w0000gn/T/.tmpYEPdVI/git-v0/checkouts/773f2287cfaa68da/a27f47c/pctasks/client
- git+https://github.com/microsoft/planetary-computer-tasks@main#subdirectory=pctasks/client
```

I just have to comment on the transitive dependencies to make it work.

**Comment by ryanc-bs** [comment]
Would it be possible to just ignore the [tool.uv.sources] section entirely in dependencies, or else allow the top level pyproject.toml to set a different package source without giving a "Requirements contain conflicting URLs" error?

This is super annoying when installing through git where you have multiple inter-dependent packages if you use git tags and/or branches to specify the version.

For example let's say package A depends on B and C, and B also depends on C. Let's say that in project B the source for C is "https://github.com/foo/bar@v1.2.3" and in project A I set the source to ""https://github.com/foo/bar@v1.2.4" instead. Currently this results in an error, so I cannot bump the version of C in A without bumping it in B first and then bumping B and C together in A. What I want instead is to just ignore the source defined in B and use A as the source of truth.

**Comment by ryanc-bs** [comment]
> Would it be possible to just ignore the [tool.uv.sources] section entirely in dependencies, or else allow the top level pyproject.toml to set a different package source without giving a "Requirements contain conflicting URLs" error?
> 
> This is super annoying when installing through git where you have multiple inter-dependent packages if you use git tags and/or branches to specify the version.
> 
> For example let's say package A depends on B and C, and B also depends on C. Let's say that in project B the source for C is "https://github.com/foo/bar@v1.2.3" and in project A I set the source to ""https://github.com/foo/bar@v1.2.4" instead. Currently this results in an error, so I cannot bump the version of C in A without bumping it in B first and then bumping B and C together in A. What I want instead is to just ignore the source defined in B and use A as the source of truth.

In summary - I'm not familiar with the uv internals but I think the solution here should be quite straightforward from a high level:

With regards to [tool.uv.sources] entries: in case of a conflict, should always use the source which is higher in the tree of dependencies. Only raise an error if two entries at the same level in the hierarchy disagree on the source.

**Comment by dnadlinger** [comment]
@ryanc-bs Agreed on the usefulness of such an override behaviour: Let's say library `A` depends on `B` which is not on PyPI and will require a Git dependency. In the public release of `A`, I want to depend on an upsteam version of `B`. For internal development, I want to have a meta-package `C` that pulls in both `A` as well as an internal fork of `B`. The end goal is to have `C`, with a committed lock file, serve as an internal base development environment, while also keeping `A` available as a public library.

For this use case, however, it seems that specifying `[tool.uv] overide-dependencies = ["B"]` in `C/pyproject.toml` is good enough (?).

**Comment by ships** [comment]
In our case, i think `override-dependencies` is sufficient to avoid multiple repo branches! Thanks for that pointer 🎉

It would be ergonomic still, to have something like `authoritative = true` or `override_transitive_sources = true` in the `tool.uv.sources` block, or allow something similar project-wide, since otherwise we would have to replicate the source info re-encoded in the python dependency string syntax, in the `override-dependencies`.

If i understood, this essentially as @ryanc-bs suggests but with a config flag rather than resolving that way by default.
